oauth2client/contrib/multiprocess_file_storage.py

```diff
@@ -0,0 +1,261 @@
+# Copyright 2016 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""File-based storage that supports multiple credentials and cross-process
+access.
+
+This module supersedes the functionality previously found in `multistore_file`.
+
+This module provides the MultiprocessFileStorage class that:
+* Is tied to a single credential via a user-specified key. This key can be used
+  to distinguish between multiple users, client ids, and/or scopes.
+* Can be safely accessed and refreshed across threads and processes.
+
+Process & thread safety guarantees the following behavior:
+* If one process refreshes a credential, subsequent refreshes from other
+  processes will re-fetch the credentials from the file instead of performing
+  an http request.
+* If two processes attempt to refresh concurrently, only one process will be
+  able to acquire the lock and refresh, with the deadlock caveat below.
+* The interprocess lock will not deadlock, instead, the if a process can not
+  acquire the interprocess lock within INTERPROCESS_LOCK_DEADLINE it will
+  allow refreshing the credential but will not write the updated credential to
+  disk, This logic happens during every lock cycle - if the credentials are
+  refreshed again it will retry locking and writing as normal.
+
+"""
+
+import base64
+import json
+import logging
+import os
+import threading
+
+import fasteners
+from six import iteritems
+
+from oauth2client.client import Credentials
+from oauth2client.client import Storage as BaseStorage
+
+INTERPROCESS_LOCK_DEADLINE = 1
+logger = logging.getLogger(__name__)
+_backends = {}
+_backends_lock = threading.Lock()
+
+
+def _get_backend(filename):
+    """A helper method to get or create a backend with thread locking.
+
+    There should only be one backend per-file per-process."""
+
```

Comment:
Definitely don't read to much into the docstrings in terms of style. I'll
update all of them once you're fine with the basic implementation here.

On Sun, May 22, 2016, 11:50 AM Nathaniel Manista <notifications@github.com>
wrote:

> In oauth2client/contrib/multiprocess_file_storage.py
> <https://github.com/google/oauth2client/pull/504#discussion_r64154382>:
>
> > +from six import iteritems
> > +
> > +from oauth2client.client import Credentials
> > +from oauth2client.client import Storage as BaseStorage
> > +
> > +INTERPROCESS_LOCK_DEADLINE = 1
> > +logger = logging.getLogger(__name__)
> > +_backends = {}
> > +_backends_lock = threading.Lock()
> > +
> > +
> > +def _get_backend(filename):
> > +    """A helper method to get or create a backend with thread locking.
> > +
> > +    There should only be one backend per-file per-process."""
> > +
>
> Blank lines between behavior doc strings and behavior bodies aren't
> generally used in this codebase.
>
> To me they look silly, but "don't mix in the same codebase having them and
> not-having them" is the reason that I would expect to be persuasive.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/google/oauth2client/pull/504/files/3067a17423c662095a52c0ffb402d66688afddd7#r64154382>
>